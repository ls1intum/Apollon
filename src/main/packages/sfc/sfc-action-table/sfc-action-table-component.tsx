import React, { FunctionComponent } from 'react';
import { ThemedLine, ThemedRect } from '../../../components/theme/themedComponents';
import { SfcActionTable } from './sfc-action-table';

interface Props {
  element: SfcActionTable;
  children?: React.ReactNode;
}

/**
 * Component for rendering a sfc action table.
 * Displays a grid with horizontal and vertical lines for organizing action rows.
 */
export const SfcActionTableComponent: FunctionComponent<Props> = ({ element, children }) => {
  const numberOfLines = Math.floor(element.bounds.height / 30) - 1;
  const horizontalLines = Array.from({ length: numberOfLines }, (_, index) => (index + 1) * 30);

  return (
    <g>
      <ThemedRect width="100%" height="100%" strokeColor="none" />
      {children}

      {horizontalLines.map((y) => (
        <ThemedLine key={y} x1={0} y1={y} x2={element.bounds.width} y2={y} stroke={element.strokeColor} />
      ))}

      <ThemedLine x1={30} y1={0} x2={30} y2={element.bounds.height} stroke={element.strokeColor} />

      <ThemedRect fillColor="none" width="100%" height="100%" strokeColor={element.strokeColor} />
    </g>
  );
};
