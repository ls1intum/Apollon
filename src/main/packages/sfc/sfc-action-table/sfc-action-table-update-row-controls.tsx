import React, { useState } from 'react';
import styled from 'styled-components';
import { Button } from '../../../components/controls/button/button';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { StylePane } from '../../../components/style-pane/style-pane';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { IUMLElement } from '../../../services/uml-element/uml-element';

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

type Props = {
  row: IUMLElement;
  update: typeof UMLElementRepository.update;
  onDelete: (id: string) => () => void;
};

export const SfcActionTableUpdateRowControls = ({ row, update, onDelete }: Props) => {
  const [colorOpen, setColorOpen] = useState(false);

  const toggleColor = () => {
    setColorOpen(!colorOpen);
  };

  const handleDelete = () => {
    onDelete(row.id)();
  };

  return (
    <>
      <Flex>
        <ColorButton onClick={toggleColor} />
        <Button color="link" tabIndex={-1} onClick={() => handleDelete()}>
          <TrashIcon />
        </Button>
      </Flex>
      <div style={{ gridColumn: '1 / -1' }}>
        <StylePane open={colorOpen} element={row} onColorChange={update} fillColor textColor />
      </div>
    </>
  );
};
