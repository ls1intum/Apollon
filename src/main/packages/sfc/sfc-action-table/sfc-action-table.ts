import { SfcContainer } from '../base/sfc-container';
import { UMLElementType } from '../../uml-element-type';
import { SfcElementType } from '../index';
import { Text } from '../../../utils/svg/text';
import { UMLElement } from '../../../services/uml-element/uml-element';
import { ILayer } from '../../../services/layouter/layer';

/**
 * Represents an action table in a sfc.
 * Contains rows of actions with identifiers and descriptions.
 */
export class SfcActionTable extends SfcContainer {
  type: UMLElementType = SfcElementType.SfcActionTable;
  override minHeight = 30;

  override childWidthCalculation: (canvas: ILayer, child: UMLElement) => number = (canvas, child) => {
    const parsedValues = JSON.parse((child as UMLElement).name);
    return Text.size(canvas, parsedValues[1] ?? 0, { fontWeight: 'normal' }).width + 50;
  };
}
