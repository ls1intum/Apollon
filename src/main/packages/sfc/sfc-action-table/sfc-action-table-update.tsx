import React, { ComponentClass, FunctionComponent, useRef, useState } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { ModelState } from '../../../components/store/model-state';
import { UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { notEmpty } from '../../../utils/not-empty';
import { AsyncDispatch } from '../../../utils/actions/actions';
import { compose } from 'redux';
import { SfcActionTable } from './sfc-action-table';
import { SfcActionTableUpdateRowControls } from './sfc-action-table-update-row-controls';
import { SfcActionTableRow } from './sfc-action-table-row/sfc-action-table-row';

const Grid = styled.div`
  display: grid;
  grid-template-columns: 30px 1fr auto;
`;

interface OwnProps {
  element: SfcActionTable;
}

interface DispatchProps {
  create: typeof UMLElementRepository.create;
  update: typeof UMLElementRepository.update;
  delete: typeof UMLElementRepository.delete;
  getById: (id: string) => UMLElement | null;
}

type Props = OwnProps & DispatchProps;

const SfcActionTableUpdateComponent: FunctionComponent<Props> = ({
  element,
  create,
  update,
  delete: deleteElement,
  getById,
}) => {
  const [fieldToFocus, setFieldToFocus] = useState<Textfield<string> | null>(null);

  const newRowFirstCell = useRef<Textfield<string>>(null);
  const newRowSecondCell = useRef<Textfield<string>>(null);

  const labelRefs: (Textfield<string> | null)[] = [];

  const children = element.ownedElements.map((id) => getById(id)).filter(notEmpty);
  const rows = children.filter((child) => child instanceof SfcActionTableRow);
  const cellValues = rows.map((row) => JSON.parse(row.name));

  // Focus the field that was set to be focused
  React.useEffect(() => {
    if (fieldToFocus) {
      fieldToFocus.focus();
      setFieldToFocus(null);
    }
  }, [fieldToFocus]);

  function rename(rowId: string, rowIndex: number, colIndex: 0 | 1, cellValue: string): void {
    const otherColIndex = 1 - colIndex;
    const rowValue = [];
    rowValue[colIndex] = cellValue;
    rowValue[otherColIndex] = cellValues[rowIndex][otherColIndex] ?? '';
    update(rowId, { name: JSON.stringify(rowValue) });
  }

  function handleDelete(id: string): () => void {
    return () => deleteElement(id);
  }

  function createRow(): void {
    const firstInput = newRowFirstCell.current?.ref.current;
    const secondInput = newRowSecondCell.current?.ref.current;

    const firstCellValue = firstInput?.value ?? '';
    const secondCellValue = secondInput?.value ?? '';

    if (firstInput) {
      firstInput.value = '';
    }
    if (secondInput) {
      secondInput.value = '';
    }

    const rowValues = [firstCellValue, secondCellValue];
    const row = new SfcActionTableRow({ name: JSON.stringify(rowValues) });
    create(row, element.id);
  }

  return (
    <div>
      <section>
        <Grid>
          {rows.map((row, index) => (
            <React.Fragment key={row.id}>
              <Textfield
                id={'first' + row.id}
                key={'first' + row.id}
                ref={(ref) => (labelRefs[2 * index] = ref)}
                value={cellValues[index][0]}
                onChange={(value) => rename(row.id, index, 0, value)}
                onSubmitKeyUp={() => setFieldToFocus(labelRefs[2 * index + 1])}
              />
              <Textfield
                id={'second' + row.id}
                key={'second' + row.id}
                ref={(ref) => (labelRefs[2 * index + 1] = ref)}
                value={cellValues[index][1]}
                onChange={(value) => rename(row.id, index, 1, value)}
                onSubmitKeyUp={() =>
                  index === rows.length - 1
                    ? newRowFirstCell.current?.focus()
                    : setFieldToFocus(labelRefs[2 * index + 2])
                }
              />
              <SfcActionTableUpdateRowControls row={row as UMLElement} update={update} onDelete={handleDelete} />
            </React.Fragment>
          ))}
        </Grid>

        <Grid>
          <Textfield
            ref={newRowFirstCell}
            outline
            value=""
            onSubmitKeyUp={() => setFieldToFocus(newRowSecondCell.current)}
          />

          <Textfield
            ref={newRowSecondCell}
            outline
            value=""
            onSubmit={createRow}
            onSubmitKeyUp={(key: string, value: string) => {
              if (value) {
                setFieldToFocus(newRowFirstCell.current);
              }
            }}
          />
          <div />
        </Grid>
      </section>
    </div>
  );
};

const enhance = compose<ComponentClass<OwnProps>>(
  connect<{}, DispatchProps, OwnProps, ModelState>(null, {
    create: UMLElementRepository.create,
    update: UMLElementRepository.update,
    delete: UMLElementRepository.delete,
    getById: UMLElementRepository.getById as any as AsyncDispatch<typeof UMLElementRepository.getById>,
  }),
);

/**
 * Component for editing rows in a sfc action table.
 * Provides a grid interface for adding, editing, and deleting action rows.
 */
export const SfcActionTableUpdate = enhance(SfcActionTableUpdateComponent);
