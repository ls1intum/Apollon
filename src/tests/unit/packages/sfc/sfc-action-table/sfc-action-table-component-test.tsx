import { wrappedRender } from '../../../test-utils/render';
import * as React from 'react';
import { SfcActionTable } from '../../../../../main/packages/sfc/sfc-action-table/sfc-action-table';
import { SfcActionTableComponent } from '../../../../../main/packages/sfc/sfc-action-table/sfc-action-table-component';

it('render the sfc-action-table-component', () => {
  const actionTable: SfcActionTable = new SfcActionTable();
  const { baseElement } = wrappedRender(
    <svg>
      <SfcActionTableComponent element={actionTable} />
    </svg>,
  );
  expect(baseElement).toMatchSnapshot();
});