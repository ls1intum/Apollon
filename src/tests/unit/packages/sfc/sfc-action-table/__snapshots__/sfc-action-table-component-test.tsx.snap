// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the sfc-action-table-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <rect
          class="sc-fqkvVR lnbwZS"
          fill="white"
          height="100%"
          stroke="none"
          width="100%"
        />
        <line
          class="sc-imWYAI loIUkN"
          stroke="black"
          x1="0"
          x2="160"
          y1="30"
          y2="30"
        />
        <line
          class="sc-imWYAI loIUkN"
          stroke="black"
          x1="0"
          x2="160"
          y1="60"
          y2="60"
        />
        <line
          class="sc-imWYAI loIUkN"
          stroke="black"
          x1="30"
          x2="30"
          y1="0"
          y2="100"
        />
        <rect
          class="sc-fqkvVR jRtNsY"
          fill="none"
          height="100%"
          stroke="black"
          width="100%"
        />
      </g>
    </svg>
  </div>
</body>
`;
